### PR TITLE
Added stream wrapper

### PR DESCRIPTION
**Pull request by bezoerb**
This PR adds a stream wrapper to simplify the usage of critical with multiple src files in gulp (#38)

### CODE REVIEW

**Comment by bezoerb** [comment]
I would prefer some review by any gulp pro in here before merging.
/cc @sindresorhus? ;)

**Comment by alanontheweb** [comment]
Hi, I had initially tried the original Critical script with success for my index page and was looking for this feature to critical different templates. I got the script to compile but I'm not sure if I'm doing something wrong. Here's what I noticed:

1) Files output with a .css extension (but contain markup and inline styles)
2)After renaming to .html, the page renders fine BUT if you remove the loadCSS script, then the page no longer renders. A quick view at the inspector tools showed me this - 

body, html {
  height: :after,:before{box-sizing:border-box}body,html{font-size:10.hide{display:none!important;visibility:hidden}.row{width:.row:before{content:" ";display:table}.row:after{clear:both}.row .row{width:auto;margin-left:-.9375rem; etc

As you can see the output is screwy. Did I do something wrong?

**Comment by bezoerb** [comment]
Hey @alanontheweb,
thanks for testing :) 
1) The wrong file extension should be fixed with my last commit. I've added some more tests to cover this behavior. 
2) Just tried this with [Web-Starter-Kit](https://developers.google.com/web/starter-kit/) and everything seems to work as expected. Only the "below the fold" content is rendered without css after removing the loadCSS block. Can you provide some more infos on this?

**Comment by alanontheweb** [comment]
Hey @bezoerb,

It's no problem. Thanks for writing this :)

So upon further testing on a clean Foundation template, it seems to be working well. On my current project, something is getting screwed up, so I'll have to debug my gulp tasks, html, css to see if I can spot something.

Did you push your last commit? I downloaded the source files from here just a few hours ago and they still have the .css bug.

**Comment by bezoerb** [comment]
Hey @alanontheweb,
i pushed my last commit just before writing the comment.
https://github.com/addyosmani/critical/pull/59/files#diff-c4506f1d6ae9e2220d7195c3c555d8b5R316

i think you'll have to redownload the source ;)

**Comment by alanontheweb** [comment]
Yup, that fixed the extension issue, haha. I'll report back if I figure out what's wrong with my other build. Thanks again!

**Comment by alanontheweb** [comment]
I got everything working now (the combineMQ gulp plugin was causing issues for some reason). Do you know if there's a way to exclude a css file from being picked up? It's adding an external font stylesheet to the <noscript> and removing it from the source. I tried specifying the CSS (  css: ['dist/styles/main.css'],) but it still doesn't seem to do anything differently.

**Comment by alanontheweb** [comment]
This is so close to working! The only two issues I'm experiencing is the one I listed above, AND that I can't use absolute paths. Example - 

Gulp is globbing through several HTML files, and some are in subdirectories. In my HTML, my path is "/assets/css/app.min.css" (note the slash at the beginning).

This does not work or result in generating multiple CSS files, and it doesn't append the appropriate filename to the LoadCSS snippet functions.

Removing the front slash to just "assets/css/app.min.css" MAKES THE WHOLE THING WORK. However, once the files are outputted, the loadCSS is pointing to a relative asset, and this breaks the pages that are inside a subdirectory, since it can't find the CSS there.

I hope this is an easy fix.

**Comment by bezoerb** [comment]
Hey @alanontheweb 
specifying the stylesheet in the css option only affects the css files used to create the critical path css.
Currently there's no option to also ignore specific styles from async loading. Note that all stylesheets which are rendered to the noscript block are also loaded async with loadCSS. 

Now the second issue. I don't see the problem yet why you can't use absolute paths, sorry.
I tried a similar scenario with the resources from wsk again. 
1. Adding some nested folders e.g. app/test/another-test/ 
2. Put styleguide.html inside 
3. Point stylesheets to absolute path `/styles/components.css`
4. run `gulp` to build dist html
5 run `gulp critical`with the following configuration
```javascript
gulp.task('critical', function () {
    gulp.src('dist/**/*.html')
        .pipe(critical({base: 'dist/',css: ['dist/styles/components.css','dist/styles/main.css']}))
        .pipe(gulp.dest('dist'));
});
```
Navigating to `localhost:3000/test/another-test/styleguide.html` afterwards causes my external stylesheet to be loaded asynchronously from `/styles/components.css` 
The whole page looks as expected.

**Comment by alanontheweb** [comment]
Okay, so I think I figured it out. It does work with both absolute and relative paths, as long as you don't use the extract function (which creates its own unique stylesheet per page). If you use  extract: true, then it only works with a relative path.

Help me understand something though - If we don't use extract, the stylesheet remains intact. When we use cssLoad to pull in the stylesheet, is it pulling in duplicate styles then (additional kb), or does the script know to ignore the styles contained in the critical CSS.

If it's pulling in the entire stylesheet w/ duplicate styles I would really like to use the extract function with absolute paths. 

gulp.task('critical', function () {
    return gulp.src('build/**/*.html')
    .pipe(critical({
        base: 'build',
        width: 1024,
         height: 768,
        minify:false,
         extract: true
    }))
    .pipe($.rename({ suffix: '.critical' }))
        .pipe(gulp.dest('build'));
});

**Comment by bezoerb** [comment]
hey @alanontheweb,
The entire stylesheet with duplicates get loaded when you disable the extract option but you can rely on browser cache if you don't load a different stylesheet for every page. See #39.
I think it makes sense to open a new ticket for the errors with extract:true

**Comment by alanontheweb** [comment]
Yeah, that's very true. I thought about the caching. It might be a good compromise that way actually. And I'll throw that into a new ticket. I wasn't sure if it was specific to the recent changes you made or in the original plugin.

Thanks so much!

**Comment by bezoerb** [comment]
Any review updates here?
/cc @addyosmani @sindresorhus

**Comment by bezoerb** [comment]
I got some unsatisfying behaviour when running the gulp task
```bash
❯ gulp critical
[00:06:40] Using gulpfile ~/Sites/web-starter-kit/gulpfile.js
[00:06:40] Starting 'critical'...
[00:06:40] Finished 'critical' after 7.11 ms
                                                                                                                                  
~/Sites/web-starter-kit  14s
``` 
Can anyone tell me why gulp assumes the task has finished after 7.11ms and it actually took 14s?
Where is the problem here and how can we fix it?

**Comment by sindresorhus** [comment]
@bezoerb Did you `return` the stream? That's how gulp know when it's done.

**Comment by bezoerb** [comment]
@sindresorhus I thought so. See https://github.com/addyosmani/critical/pull/59/files#diff-c4506f1d6ae9e2220d7195c3c555d8b5R295
But maybe i got something wrong. I've not completely understood the whole stream thing yet ;)

File: src/critical.js [hunk 1]
```diff
@@ -281,3 +274,51 @@ exports.generateInline = function (opts, cb) {
...
+            return this.emit('error', new PluginError('critical', 'Streaming not supported'));
+        }
+
+        var options = _.assign(opts,{
+            html: file.contents.toString()
+        });
+
+        exports[command](options, function(err,data){
+            if (err) {
+                throw new PluginError('critical', err.message);
```
Comment (sindresorhus) [inline src/critical.js]: You shouldn't throw in async. Return the error.

File: src/critical.js [hunk 2]
```diff
@@ -281,3 +274,51 @@ exports.generateInline = function (opts, cb) {
...
+            }
+
+            // rename file if not inlined
+            if (opts.inline === false) {
+                file.path = replaceExtension(file.path,'.css');
+            }
+
+            file.contents = new Buffer(data, 'utf-8');
```
Comment (sindresorhus) [inline src/critical.js]: encoding is moot here

File: src/critical.js [hunk 3]
```diff
@@ -281,3 +274,51 @@ exports.generateInline = function (opts, cb) {
...
+        });
+
+        exports[command](options, function(err,data){
+            if (err) {
+                throw new PluginError('critical', err.message);
+            }
+
+            // rename file if not inlined
+            if (opts.inline === false) {
+                file.path = replaceExtension(file.path,'.css');
```
Comment (sindresorhus) [inline src/critical.js]: i think the dot here is moot

**Comment by bezoerb** [comment]
@sindresorhus thanks!

File: src/critical.js (hunk 3 above)
Comment (bezoerb) [inline src/critical.js]: it would result in `somefilecss` if i'll drop it

Comment (sindresorhus) [inline src/critical.js]: oh, nvm.

**Comment by sindresorhus** [comment]
@bezoerb You need to fix the merge conflict.

**Comment by bezoerb** [comment]
@sindresorhus merge conflicts fixed locally but my tests are failing (so does travis do on `master`) 
We need to fix cli tests and add the missing file for the multidimensional css test (`fixture/test-adaptive-final.css` - `fixture/test-` is ignored
I'll go for the CLI changes (or do you want? ;)) 
@samccone could you rename your css file and add it?

When my local `npm test` is green i'll push & rebase

**Comment by samccone** [comment]
@bezoerb no problem, will do it tonight when I get back to my other computer. Odd that my branch even passed on CI... raises an interesting question as to why it ever passed

**Comment by bezoerb** [comment]
@samccone yep, that' really odd. Npm shouldn't exit with code 0 on errors.

**Comment by samccone** [comment]
xref https://github.com/addyosmani/critical/issues/67

**Comment by bezoerb** [comment]
@samccone tests are working as expected now. I skipped your test until we can add your test files. 
/cc @addyosmani @sindresorhus

**Comment by samccone** [comment]
@bezoerb looks like you need to rebase off of the latest master so that this can be cleanly merged

**Comment by bezoerb** [comment]
this is just a few seconds ahead ^^

**Comment by addyosmani** [comment]
Correct me if I'm wrong: we're now clear on tests passing, but CLI exit codes are still giving us issues. This is otherwise good to go soon.

@samccone Thanks for stepping in to help oversee some of the reviews. It's appreciated.

**Comment by bezoerb** [comment]
@addyosmani you're right. I already got a fix for the exit codes and i'm currently trying to write a testcase for it

**Comment by micahblu** [comment]
I'm getting an error trying to implement this. Also it might be a good idea to put up example usage in the readme. 

I have the following gulpfile.js
```js

var gulp = require('gulp'),
    critical = require('critical'),
    fs = require('fs');

gulp.task('critical', function () {
  gulp.src('dist/**/*.html')
    .pipe(critical({
        base: 'dist/',
        css: ['dist/stylesheets/styles.css']
    }))
    .pipe(gulp.dest('dist'));
});
```

and I'm getting the following error:
```shell
[15:14:09] Starting 'critical'...
[15:14:09] 'critical' errored after 6.95 ms
[15:14:09] TypeError: object is not a function
    at Gulp.<anonymous> (/Users/micahblu/Projects/WireCash-Website/gulpfile.js:8:11)
    at module.exports (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:273:3)
    at Gulp.Orchestrator._runStep (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:214:10)
    at Gulp.Orchestrator.start (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:134:8)
    at /usr/local/lib/node_modules/gulp/bin/gulp.js:121:20
    at process._tickCallback (node.js:419:13)
    at Function.Module.runMain (module.js:499:11)
    at startup (node.js:119:16)
    at node.js:906:3
```

**Comment by bezoerb** [comment]
@micahblu you need to adjust your require statement `var critical = require('critical').stream;`
See: https://github.com/addyosmani/critical/tree/critical-gulp-stream#gulp

**Comment by micahblu** [comment]
@bezoerb thanks for the quick reply, I had noticed that I was missing the .stream after I commented, I have added it, unfortunately the error persists.
 
gulpfile is now:
```js 
var gulp = require('gulp'),
    critical = require('critical').stream,
    fs = require('fs');


// Generate & Inline Critical-path CSS
gulp.task('critical', function () {
    return gulp.src('dist/**/*.html')
        .pipe(critical({base: 'dist/',css: ['dist/stylesheets/styles.css']}))
        .pipe(gulp.dest('dist'));
});
```

error is:
```shell
[15:25:28] Starting 'critical'...
[15:25:28] 'critical' errored after 6.12 ms
[15:25:28] TypeError: undefined is not a function
    at Gulp.<anonymous> (/Users/micahblu/Projects/WireCash-Website/gulpfile.js:10:15)
    at module.exports (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:273:3)
    at Gulp.Orchestrator._runStep (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:214:10)
    at Gulp.Orchestrator.start (/Users/micahblu/Projects/WireCash-Website/node_modules/gulp/node_modules/orchestrator/index.js:134:8)
    at /usr/local/lib/node_modules/gulp/bin/gulp.js:121:20
    at process._tickCallback (node.js:419:13)
    at Function.Module.runMain (module.js:499:11)
    at startup (node.js:119:16)
    at node.js:906:3
```

**Comment by micahblu** [comment]
I have found that critical is undefined when .stream is added. I'm not that familiar with node streams so I'm unsure why that is happening. When I remove the .stream the critical object is restored

**Comment by micahblu** [comment]
@bezoerb looks like the stream export is missing in index.js?

**Comment by bezoerb** [comment]
This one is not merged yet. Are you shure you are using the `critical-gulp-stream` branch?
`npm install --save-dev git://github.com/addyosmani/critical#critical-gulp-stream`

**Comment by micahblu** [comment]
I am not! lol that would explain it.. +10 to get this merged to master asap. Thanks @bezoerb :)

Works, though I am getting some warnings:

```shell
node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at process.addListener (events.js:160:15)
    at process.on.process.addListener (node.js:773:26)
    at module.exports (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/penthouse/lib/index.js:69:13)
    at tryCatcher (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/util.js:24:31)
    at ret (eval at <anonymous> (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/promisify.js:154:12), <anonymous>:12:39)
    at /Users/micahblu/Projects/WireCash-Website/node_modules/critical/index.js:167:20
    at tryCatcher (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/util.js:24:31)
    at Promise._settlePromiseFromHandler (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/promise.js:454:31)
    at Promise._settlePromiseAt (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/promise.js:530:18)
    at Async._drainQueue (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/async.js:84:12)
    at Async._drainQueues (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/async.js:89:10)
    at Async.drainQueues (/Users/micahblu/Projects/WireCash-Website/node_modules/critical/node_modules/bluebird/js/main/async.js:14:14)
    at process._tickCallback (node.js:419:13)
```

**Comment by bezoerb** [comment]
Played around with this branch a bit and everything seems to work just as expected.
From my perspective this could be merged for the next release

**Comment by addyosmani** [comment]
> @addyosmani would you mind adding the appveyor hooks mentioned in #45?

This should be enabled now. If we've missed anything at all, feel free to yell!

**Comment by danimalweb** [comment]
Seems to be working well. However I am also getting the memory leak warnings.

```(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at process.addListener (events.js:160:15)
    at process.on.process.addListener (node.js:802:26)
    at module.exports (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/penthouse/lib/index.js:69:13)
    at tryCatcher (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/util.js:24:31)
    at ret (eval at <anonymous> (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/promisify.js:154:12), <anonymous>:12:39)
    at /home/admin/web/lambda.local/public_html/node_modules/critical/index.js:167:20
    at tryCatcher (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/util.js:24:31)
    at Promise._settlePromiseFromHandler (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/promise.js:454:31)
    at Promise._settlePromiseAt (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/promise.js:530:18)
    at Async._drainQueue (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/async.js:182:12)
    at Async._drainQueues (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/async.js:187:10)
    at Async.drainQueues (/home/admin/web/lambda.local/public_html/node_modules/critical/node_modules/bluebird/js/main/async.js:15:14)
    at process._tickCallback (node.js:448:13)```

**Comment by zgosalvez** [comment]
Installed this today. Been getting the same `TypeError: undefined is not a function` error, after following the instructions at https://github.com/addyosmani/critical#gulp 

If I remove `.stream` I get `TypeError: object is not a function`

**Comment by bezoerb** [comment]
Try to install master from npm npm install git+https://github.com/addyosmani/critical.git as this feature hasn't been shipped yet.

**Comment by zgosalvez** [comment]
@bezoerb Oh ok. I assumed it's already available since it's in master.

**Comment by Snugug** [comment]
I'm likewise getting memory leak warnings

**Pull request merged**